### ISSUE TITLE
proot vs Android 5.1.1

### ISSUE THREAD

**Issue by davidgiven** [corbinlc/GNURootDebian#14]
On my 5.1.1 Nexus 7, it looks like there are still issues with proot.

Unfortunately strace hangs, so I can't get a system call log, but:

$ touch a/b/c/d
touch: cannot touch 'a/b/c/d': Bad address

This looks like the same issue the old version had, which is that the wrong errno is being returned and confusing things.

In addition, there still seems to be problems with uid translation:

# su dg
$ sudo ls
sudo: /usr/bin/sudo must be owned by uid 0 and have the setuid bit set
$ ls -l /usr/bin/sudo
-rwsr-xr-x. 1 dg dg 106820 Mar 12 03:35 /usr/bin/sudo
$ exit
# ls -l /usr/bin/sudo
-rwsr-xr-x. 1 root root 106820 Mar 12 03:35 /usr/bin/sudo

**Comment by unforgettableid** [comment]
Hi!  Thanks for your report.

> strace hangs, so I can't get a system call log

Is it possible that this is a bug in strace?  If so, have you reported it to the strace maintainer?

Does ltrace work?

I'm not a GNURoot Debian developer.  Regarding the rest of what you wrote, I'll leave it for someone else to reply to.

What version of GNURoot Debian are you running?  Every time you report a bug, please tell us your GNURoot Debian version number.

**Comment by unforgettableid** [comment]
Hmmm.  Strace and ltrace each [use ptrace](https://github.com/corbinlc/gnuroot/issues/55#issuecomment-98473961), and GNURoot Debian uses PRoot which also uses ptrace.

Maybe this is why strace and ltrace don't work inside GNURoot Debian.

**Comment by unforgettableid** [comment]
By the way:  In the future, please open up a separate ticket for each problem you encounter.  Thank you!

**Comment by davidgiven** [comment]
Yeah, I was totally expecting strace to fail. And these are in the same report because they're both proot doing the wrong thing, and are very likely related...

This is version 0.2.6_armhf of GNURoot. I don't know which version of proot it is.

Do you know if there's a way to persuade GNURoot to use an external proot binary so I can try building my own?

**Comment by unforgettableid** [comment]
> these are in the same report because they're both proot doing the wrong thing, and are very likely related...

I didn't realize. My mistake.

> This is version 0.2.6_armhf of GNURoot. I don't know which version of proot it is.

OK; thanks.

As for which proot version it is:  You didn't have to tell us.  But I think it may be a version of proot which Corbin customized and built himself.  See [here](https://github.com/corbinlc/GNURootDebian/issues/11).

> Do you know if there's a way to persuade GNURoot to use an external proot binary so I can try building my own?

There may be other ways; but here's one way which I think will work.

1. Custom-build your own `proot` binary. To avoid confusion later, change its version number string to something like "5.1-dg1" so that you'll know it's one which you built yourself.
1. Root your phone.
1. Look in `/data/data/com.gnuroot.debian/support`. You'll probably find a `proot` binary there.
1. If you find one, replace it with one which you've custom-built.
1. Kill and restart GNURoot.

Does this work for you?

**Comment by corbinlc** [comment]
As far as getting the wrong error code.  There are many examples like this.  I now have a test sweet for detecting such things, but haven't had time to run around fixing them.  

As far as su and sudo.  These will be faked out in the future such that they are aliases that impact how the fake_root PRoot extension behave.  Then they will always succeed and programs will think you are the correct user.  The way things work is you are always the same users, we are just faking out what user it looks like you are as far as programs running inside the PRooted rootfs are concerned.

Currently, you start up as a fake root user and can su to any non-root user that you have created.  In the future, you will start as a non-root user and the changes to the fake_root extension will allow you to become a fake root user or anything else for that matter.

As far as strace and PRoot, I believe there is a discussion of this on the PRoot forum about this.  I think maybe you can use it inside PRoot if you build off of the latest version, or maybe it was planned work.  I don't remember.  

The source code for the PRoot version I am using can be found here: 
https://github.com/corbinlc/GNURootDebian/tree/master/GNURootDebianSource/src/main/build_rootfs/PRoot

The /support directory is accessible from within the rootfs, so you can copy a file over there as you wish.  Or you can copy a new version and then change the launchProot script (also found in the /support directory) .

Really everything should be up on github now, so you can rebuild the whole APK if you desire.  This would allow you to replace the proot exectuable and the launchProot script as you see fit.

If you can work these issues...great!  If not, I will take them and fix them overtime.

**Comment by davidgiven** [comment]
I don't particularly care about strace, but the errno issue is basically stopping everything from working. I'm glad there's a fix in sight.

My device is, unfortunately, unrootable, which is why I'm using proot. Luckily I have a manual setup I was using proot with before I unwisely upgraded to Android 5.1. I'll have a go at building the latest proot from github, and see if that works there. (But if you know where I can get a recent precompiled binary, that'd be awesome.)

**Comment by AdamMatthewLTM** [comment]
Hi all,

I am currently assisting the developers in sorting relevant issues for GNURoot Debian on GitHub. If the issue discussed is still relevant, could you please reply to this thread and state the issue needs to be resolved and I will attempt to sort it for you ASAP. If the issue is no longer relevant, please reply to this thread and state it has been resolved, and I will suggest this thread be closed.

Furthermore, if a particular issue is considered to be high priority, please state that you are requesting it to be considered as high priority and I will mark it as such.

Thank you all so much for your time and understanding,
-Adam
